### ISSUE TITLE
[windows] [error] [storage] [cio file] too many open chunks | storage.max_chunks_up ignored

### ISSUE THREAD

**Issue by FaustUser** [fluent/fluent-bit#3338]
## Bug Report

**Describe the bug**
Fluent-bit parses NPS log files by regular expression into a json document and sends the result to kafka. When the application is running and memory usage grows to storage.max_chunks_up * 2MB (as the documentation says), an error is constantly being written to the log:
```
[2021/04/07 09:46:29] [error] [storage] [cio file] too many open chunks on 'tail.0/3988-1617777986.989031600.flb' (cio_file_up() line=791)
[2021/04/07 09:46:29] [error] [storage] [cio file] too many open chunks on 'tail.0/3988-1617777988.741708500.flb' (cio_file_up() line=791)
[2021/04/07 09:46:34] [error] [storage] [cio file] too many open chunks on 'tail.0/3988-1617777990.767973800.flb' (cio_file_up() line=791)
[2021/04/07 09:46:34] [error] [storage] [cio file] too many open chunks on 'tail.0/3988-1617777990.972639700.flb' (cio_file_up() line=791)
[2021/04/07 09:46:34] [error] [storage] [cio file] too many open chunks on 'tail.0/3988-1617777994.643476500.flb' (cio_file_up() line=791)
[2021/04/07 09:46:38] [error] [storage] [cio file] too many open chunks on 'tail.0/3988-1617777990.972639700.flb' (cio_file_up() line=791)
[2021/04/07 09:46:38] [error] [storage] [cio file] too many open chunks on 'tail.0/3988-1617777994.643476500.flb' (cio_file_up() line=791)
[2021/04/07 09:46:38] [error] [storage] [cio file] too many open chunks on 'tail.0/3988-1617777995.992295200.flb' (cio_file_up() line=791)
[2021/04/07 09:46:38] [error] [storage] [cio file] too many open chunks on 'tail.0/3988-1617777996.494183600.flb' (cio_file_up() line=791)
[2021/04/07 09:46:43] [error] [storage] [cio file] too many open chunks on 'tail.0/3988-1617777990.972639700.flb' (cio_file_up() line=791)
[2021/04/07 09:46:43] [error] [storage] [cio file] too many open chunks on 'tail.0/3988-1617777994.643476500.flb' (cio_file_up() line=791)
[2021/04/07 09:46:43] [error] [storage] [cio file] too many open chunks on 'tail.0/3988-1617777995.992295200.flb' (cio_file_up() line=791)
[2021/04/07 09:46:43] [error] [storage] [cio file] too many open chunks on 'tail.0/3988-1617777996.494183600.flb' (cio_file_up() line=791)
[2021/04/07 09:46:43] [error] [storage] [cio file] too many open chunks on 'tail.0/3988-1617777999.398899200.flb' (cio_file_up() line=791)
```
**Memory usage continues to grow up to 600+ MB. Then file processing ends and memory usage drops to ~ 20 MB. storage.max_chunks_up ignored.**
Total files size 1GB
For tests, new data is not written to files.
All lines from log files fall into kafka without loss.

**To Reproduce**
- Steps to reproduce the problem:
  - Create Directory with NPS logs.
    Size of the directory with logs ~ 1 GB, the number of files 42, the size of each file ~ 25MB, the number of log lines in each file ~ 66500, the total number of log lines - 2742177.
  - NPS log message example:
  ```
  "Test-Test-01","IAS",03/16/2021,06:28:42,4,"Test\user",,"00-50-56-80-72-59","74-e5-f9-d3-5c-98",,"172.172.172.172","Test-Test-285-3","10.10.10.10",1,0,"10.10.10.10","test-test",03/16/2021 03:28:42,,19,,,,,,0,"311 1 10.89.87.20 03/15/2021 23:03:19 7665",,,,,3,0,1318979,1291927,"6050259b/74:e5:f9:d3:5c:98/49525718",1,72,6042,3402,,,,,,,,,,,,,,,,,,,,,"Use Windows authentication for all users",,,,,
  ```
  - Remove fluent-bit state and log files.
  - Run fluent-bit.

**Your Environment**

* Version used:
fluent-bit 1.7.3
* Operating System and version:
Microsoft Windows Server 2012 R2 x64 Standard 6.3.9600 N/A Build 9600
4 Processors installed - Intel64 Family 6 Model 63 Stepping 0 GenuineIntel ~2297 Mhz
Total physical memory 8192 MB
* Filters and plugins:
[INPUT] - tail
[FILTER] - rewrite_tag
[OUTPUT] - kafka

**Additional context**
* Fluent-bit config file:
```
[SERVICE]
    flush        5
    grace        5
    daemon       off
    log_file C:\fluent-bit\log.log
    log_level    warning
    parsers_file parsers.conf
    plugins_file plugins.conf
    http_server off
    http_listen 0.0.0.0
    http_port 2020
    storage.path C:\fluent-bit\storage\
    storage.sync normal
    storage.checksum off
    storage.max_chunks_up 128
    storage.backlog.mem_limit 5M
    storage.metrics off

[INPUT]
    name tail
    tag radius
    path C:\NPSLOGS\*.log
    read_from_head true
    db C:\fluent-bit\state.db
    storage.type filesystem
    mem_buf_limit 256M
    parser radius

[FILTER]
    name rewrite_tag
    match radius
    rule $log ^.*$ not_parsed false

[OUTPUT]
    name kafka
    match radius
    brokers *.*.*.*:9092
    topics radius
    format json
    rdkafka.enable.idempotence true
    rdkafka.security.protocol sasl_plaintext
    rdkafka.sasl.mechanisms PLAIN
    rdkafka.sasl.username username_changed
    rdkafka.sasl.password password_changed

[OUTPUT]
    name kafka
    match not_parsed
    brokers *.*.*.*:9092
    topics notParsedRadius
    format json
    rdkafka.enable.idempotence true
    rdkafka.security.protocol sasl_plaintext
    rdkafka.sasl.mechanisms PLAIN
    rdkafka.sasl.username username_changed
    rdkafka.sasl.password password_changed
```
* Fluent-bit parser file:
```
[PARSER]
    Name    radius
    Format  regex
    Regex   ^(?:"(?<computerName>.*?)")?,(?:"(?<serviceName>.*?)")?,(?<recordDate>(?:0?[1-9]|1[0-2])[\/-](?:(?:0[1-9])|(?:[12][0-9])|(?:3[01])|[1-9])[\/-](?>\d\d){1,2})?,(?<recordTime>(?:2[0123]|[01]?[0-9]):(?:[0-5][0-9]):(?:[0-5][0-9]|60))?,(?<packetType>[1-4])?,(?:"(?<userName>.*?)")?,(?:"(?<fullyQualifiedDistinguishedName>.*?)")?,(?:"(?<calledStationId>.*?)")?,(?:"(?<callingStationId>.*?)")?,(?:"(?<callbackNumber>.*?)")?,(?:"(?<framedIpAddress>.*?)")?,(?:"(?<nasIdentifier>.*?)")?,(?:"(?<nasIpAddress>.*?)")?,(?<nasPort>\d+)?,(?<clientVendor>\d+)?,(?:"(?<clientIpAddress>.*?)")?,(?:"(?<clientFriendlyName>.*?)")?,(?<eventTimestamp>(?:0?[1-9]|1[0-2])[\/-](?:(?:0[1-9])|(?:[12][0-9])|(?:3[01])|[1-9])[\/-](?>\d\d){1,2}[- ](?:2[0123]|[01]?[0-9]):(?:[0-5][0-9]):(?:[0-5][0-9]|60))?,(?<portLimit>\d+)?,(?<nasPortType>\d+)?,(?:"(?<connectInfo>.*?)")?,(?<framedProtocol>\d+)?,(?<serviceType>\d+)?,(?<authenticationType>[1-8])?,(?:"(?<policyName>.*?)")?,(?<reasonCode>[0-8]|1[6-8]|3[2-8]|48|6[4-9]|7[0-1]|80|9[6-7])?,(?:"(?<class>.*?)")?,(?<sessionTimeout>\d+)?,(?<idleTimeout>\d+)?,(?<terminationAction>\d+)?,(?:"(?<eapFriendlyName>.*?)")?,(?<acctStatusType>\d+)?,(?<acctDelayTime>\d+)?,(?<acctInputOctets>\d+)?,(?<acctOutputOctets>\d+)?,(?:"(?<acctSessionId>.*?)")?,(?<acctAuthentic>\d+)?,(?<acctSessionTime>\d+)?,(?<acctInputPackets>\d+)?,(?<acctOutputPackets>\d+)?,(?<acctTerminateCause>\d+)?,(?:"(?<acctMultiSsnId>.*?)")?,(?<acctLinkCount>\d+)?,(?<acctInterimInterval>\d+)?,(?<tunnelType>\d+)?,(?<tunnelMediumType>\d+)?,(?:"(?<tunnelClientEndpt>.*?)")?,(?:"(?<tunnelServerEndpt>.*?)")?,(?:"(?<acctTunnelConn>.*?)")?,(?:"(?<tunnelPvtGroupId>.*?)")?,(?:"(?<tunnelAssignmentId>.*?)")?,(?<tunnelPreference>\d+)?,(?<msAcctAuthType>\d+)?,(?<msAcctEapType>\d+)?,(?:"(?<msRasVersion>.*?)")?,(?<msRasVendor>\d+)?,(?:"(?<msChapError>.*?)")?,(?:"(?<msChapDomain>.*?)")?,(?<msMppeEncryptionTypes>\d+)?,(?<msMppeEncryptionPolicy>\d+)?,(?:"(?<proxyPolicyName>.*?)")?,(?<providerType>[0-2])?,(?:"(?<providerName>.*?)")?,(?:"(?<remoteServerAddress>.*?)")?,(?:"(?<msRasClientName>.*?)")?,(?<msRasClientVersion>\d+)?$
    Types packetType:integer nasPort:integer clientVendor:integer portLimit:integer nasPortType:integer framedProtocol:integer serviceType:integer authenticationType:integer reasonCode:integer sessionTimeout:integer idleTimeout:integer terminationAction:integer acctStatusType:integer acctDelayTime:integer acctInputOctets:integer acctOutputOctets:integer acctAuthentic:integer acctSessionTime:integer acctInputPackets:integer acctOutputPackets:integer acctTerminateCause:integer acctLinkCount:integer acctInterimInterval:integer tunnelType:integer tunnelMediumType:integer tunnelPreference:integer msAcctAuthType:integer msAcctEapType:integer msRasVendor:integer msMppeEncryptionTypes:integer msMppeEncryptionPolicy:integer providerType:integer msRasClientVersion:integer

```

**Comment by FaustUser** [comment]
Also there is a problem when sending data to kafka. There are 20 partitions in the topic, but the data falls into only 10 and unevenly. The spread in the number of messages is huge. But I don't know if it's fluent-bit or librdkafka problem. I would appreciate it if you check this and advise a solution.

**Comment by fujimotos** [comment]
> [2021/04/07 09:46:29] [error] [storage] [cio file] too many open chunks on 'tail.0/3988-1617777986.989031600.flb'

@FaustUser  This error message _per se_ is actually a non-problem. It indicates that
Fluent Bit honored `storage.max_chunks_up` and stopped creating new chunks to keep
quota for chunk counts. Nothing was ignored here.

That said, I agree with you that this error message is confusing. I have just posted a
quick fix edsiper/chunkio/pull/66 to make the logging less confusing.

> Memory usage continues to grow up to 600+ MB.

I could not reproduce this part. I parepared 42 files, each contains 66500 lines of
NPS logs, and telll Fluent Bit to send them to a Kafka cluster from WinServer 2019.

It took 5 minute to process all of the log. The maximum memory usage I saw was
<60mb. So it seems working as normal

Can you tell me what happens if you remove the `filter_rewrite` from your configuration?
(= test with one tail + one kafka). I suspect the tag-rewriter causes the memory bloat.

**Comment by FaustUser** [comment]
@fujimotos
> It took 5 minute to process all of the log. The maximum memory usage I saw was
<60mb. So it seems working as normal
 Can you tell me what happens if you remove the filter_rewrite from your configuration?
(= test with one tail + one kafka). I suspect the tag-rewriter causes the memory bloat.

It took me 5 minutes 20 seconds to process all the files.

![image](https://user-images.githubusercontent.com/21306326/114158212-1cc23600-992d-11eb-8c6c-0de43d7eb17c.png)

Here is fluent-bit.conf config file without rewrite filter and one kafka (kafka is on another machine - CentOS Linux release 7.7.1908 (Core)):
```
[SERVICE]
    flush        5
    grace        5
    daemon       off
    log_file C:\fluent-bit\log.log
    log_level    warning
    parsers_file parsers.conf
    plugins_file plugins.conf
    http_server off
    http_listen 0.0.0.0
    http_port 2020
    storage.path C:\fluent-bit\storage\
    storage.sync normal
    storage.checksum off
    storage.max_chunks_up 128
    storage.backlog.mem_limit 5M
    storage.metrics off

[INPUT]
    name tail
    tag radius
    path C:\NPSLOGS\*.log
    read_from_head true
    db C:\fluent-bit\state.db
    storage.type filesystem
    mem_buf_limit 256M
    parser radius

[OUTPUT]
    name kafka
    match radius
    brokers 192.168.10.99:29092
    topics radius
    format json
    rdkafka.enable.idempotence true
    rdkafka.security.protocol sasl_plaintext
    rdkafka.sasl.mechanisms PLAIN
    rdkafka.sasl.username kafka
    rdkafka.sasl.password 123
```
parsers.conf:
```
[PARSER]
    Name    radius
    Format  regex
    Regex   ^(?:"(?<computerName>.*?)")?,(?:"(?<serviceName>.*?)")?,(?<recordDate>(?:0?[1-9]|1[0-2])[\/-](?:(?:0[1-9])|(?:[12][0-9])|(?:3[01])|[1-9])[\/-](?>\d\d){1,2})?,(?<recordTime>(?:2[0123]|[01]?[0-9]):(?:[0-5][0-9]):(?:[0-5][0-9]|60))?,(?<packetType>[1-4])?,(?:"(?<userName>.*?)")?,(?:"(?<fullyQualifiedDistinguishedName>.*?)")?,(?:"(?<calledStationId>.*?)")?,(?:"(?<callingStationId>.*?)")?,(?:"(?<callbackNumber>.*?)")?,(?:"(?<framedIpAddress>.*?)")?,(?:"(?<nasIdentifier>.*?)")?,(?:"(?<nasIpAddress>.*?)")?,(?<nasPort>\d+)?,(?<clientVendor>\d+)?,(?:"(?<clientIpAddress>.*?)")?,(?:"(?<clientFriendlyName>.*?)")?,(?<eventTimestamp>(?:0?[1-9]|1[0-2])[\/-](?:(?:0[1-9])|(?:[12][0-9])|(?:3[01])|[1-9])[\/-](?>\d\d){1,2}[- ](?:2[0123]|[01]?[0-9]):(?:[0-5][0-9]):(?:[0-5][0-9]|60))?,(?<portLimit>\d+)?,(?<nasPortType>\d+)?,(?:"(?<connectInfo>.*?)")?,(?<framedProtocol>\d+)?,(?<serviceType>\d+)?,(?<authenticationType>[1-8])?,(?:"(?<policyName>.*?)")?,(?<reasonCode>[0-8]|1[6-8]|3[2-8]|48|6[4-9]|7[0-1]|80|9[6-7])?,(?:"(?<class>.*?)")?,(?<sessionTimeout>\d+)?,(?<idleTimeout>\d+)?,(?<terminationAction>\d+)?,(?:"(?<eapFriendlyName>.*?)")?,(?<acctStatusType>\d+)?,(?<acctDelayTime>\d+)?,(?<acctInputOctets>\d+)?,(?<acctOutputOctets>\d+)?,(?:"(?<acctSessionId>.*?)")?,(?<acctAuthentic>\d+)?,(?<acctSessionTime>\d+)?,(?<acctInputPackets>\d+)?,(?<acctOutputPackets>\d+)?,(?<acctTerminateCause>\d+)?,(?:"(?<acctMultiSsnId>.*?)")?,(?<acctLinkCount>\d+)?,(?<acctInterimInterval>\d+)?,(?<tunnelType>\d+)?,(?<tunnelMediumType>\d+)?,(?:"(?<tunnelClientEndpt>.*?)")?,(?:"(?<tunnelServerEndpt>.*?)")?,(?:"(?<acctTunnelConn>.*?)")?,(?:"(?<tunnelPvtGroupId>.*?)")?,(?:"(?<tunnelAssignmentId>.*?)")?,(?<tunnelPreference>\d+)?,(?<msAcctAuthType>\d+)?,(?<msAcctEapType>\d+)?,(?:"(?<msRasVersion>.*?)")?,(?<msRasVendor>\d+)?,(?:"(?<msChapError>.*?)")?,(?:"(?<msChapDomain>.*?)")?,(?<msMppeEncryptionTypes>\d+)?,(?<msMppeEncryptionPolicy>\d+)?,(?:"(?<proxyPolicyName>.*?)")?,(?<providerType>[0-2])?,(?:"(?<providerName>.*?)")?,(?:"(?<remoteServerAddress>.*?)")?,(?:"(?<msRasClientName>.*?)")?,(?<msRasClientVersion>\d+)?$
    Types packetType:integer nasPort:integer clientVendor:integer portLimit:integer nasPortType:integer framedProtocol:integer serviceType:integer authenticationType:integer reasonCode:integer sessionTimeout:integer idleTimeout:integer terminationAction:integer acctStatusType:integer acctDelayTime:integer acctInputOctets:integer acctOutputOctets:integer acctAuthentic:integer acctSessionTime:integer acctInputPackets:integer acctOutputPackets:integer acctTerminateCause:integer acctLinkCount:integer acctInterimInterval:integer tunnelType:integer tunnelMediumType:integer tunnelPreference:integer msAcctAuthType:integer msAcctEapType:integer msRasVendor:integer msMppeEncryptionTypes:integer msMppeEncryptionPolicy:integer providerType:integer msRasClientVersion:integer
```
NPS logs directory:
![image](https://user-images.githubusercontent.com/21306326/114159372-64958d00-992e-11eb-9a62-6e8cd3dca2fd.png)
![image](https://user-images.githubusercontent.com/21306326/114159437-77a85d00-992e-11eb-80a9-f244919e1d7b.png)

**Comment by josephflynch** [comment]
This is affecting GCP customer(s).

**Comment by qingling128** [comment]
We observed the same issue with 1.7.6, but it was fixed by 1.7.8. Please give that a try.
